week-2-wide-perspective/w02e02-data-manipulation/src/observers/ordersObserver.js

```diff
@@ -0,0 +1,80 @@
+import { addOrder, getOrders } from "../services/http.service.js";
+
+export class Observable {
+    constructor() {
+        this.observers = [];
+
+    }
+    addObserver(observer) {
+        this.observers.push(observer);
+    }
+
+
+    removeObserver(observer) {
+        const removeIndex = this.observers.indexOf(observer);
+        if (removeIndex !== -1) {
+            this.observers = this.observers.slice(removeIndex, 1);
+        }
+    }
+
+
+    notify(data) {
+        console.log('notify')
+        if (this.observers.length > 0) {
+            this.observers.forEach((observer) => observer.update(data))
+        }
+    }
+}
+
+export class Observer {
+    // Gets called by the Subject::notify method.
+    update() { }
+}
+export class DataTile extends Observer {
+    constructor(componentToUpdate, dataToDisplay) {
+        super()
+        this.componentToUpdate = componentToUpdate
+        this.dataToDisplay = dataToDisplay
+    }
+    render(state) {
+        this.componentToUpdate.innerHTML = state[this.dataToDisplay]
+    }
+    update(state) {
+        this.render(state)
+    }
+
+}
+
+export class State extends Observable {
+    orders = []
+    constructor() {
+        super()
+
+        // this.orders = getOrders()
+        this.state = {}
+    }
+    calculateProperties() {
+        const allAvailableOrders = this.orders
+        this.state.totalOrders = allAvailableOrders.length
+        this.state.totalIncome = allAvailableOrders.reduce((total, item) => total + item.sale, 0).toFixed(2)
+        this.state.bestSale = allAvailableOrders.sort((a, b) => b.sale - a.sale)[0].sale
+    }
+    addNewOrder(order) {
```

Comment:
generalnie w programie jest charakterystyczny `bug` że czasem odświeża się licznik co `2`gie kliknięcie. To jest jego przyczyną.

```javascript
addOrder(order)
```

jest asynchroniczne, i nie możesz puścić `updateOrders` zanim się nie wykona.
Masz tutaj `racing condition` chcesz zrobić `updateOrders` w tym samym momencie co `addOrder`

powinno być:
```javascript
async addNewOrder(order) {
        await addOrder(order)
        await this.updateOrders()
    }
```

wtedy działa jak złoto.